File: modules/x-engine-module-core/iOS/Class/iOS_Modules/core/GlobalState.m
//
//  GlobalState.m
//  ModuleApp
//
//  Created by zk on 2021/3/27.
//  Copyright © 2021 zkty-team. All rights reserved.
//

#import "GlobalState.h"
#import "HistoryModel.h"
@interface GlobalState()
@property(nonatomic,strong) NSMapTable<id,NSMutableArray<HistoryModel*>*>* wv__vc_paths;

@end
@implementation GlobalState
static NSString*  s_microapp_root_url;
static XEngineWebView*  s_showing_webview;


- (instancetype)init {
   self = [super init];
   self.wv__vc_paths = [NSMapTable new];
   return self;
}


+ (instancetype)sharedInstance {
    static GlobalState *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[super allocWithZone:NULL] init];

    });
    return sharedInstance;
}

+ (NSString * _Nullable) s_microapp_root_url{
    return s_microapp_root_url;
}
+ (void) set_s_microapp_root_url:(NSString*)val{
      s_microapp_root_url=val;
}
+ (void)setCurrentWebView:(XEngineWebView*) val{
    if(s_showing_webview)
        s_showing_webview=nil;
    s_showing_webview=val;
}
+ (XEngineWebView*)getCurrentWebView{
    return s_showing_webview;
}
- (void)deleteWebView:(XEngineWebView *) webview{
}
- (NSMutableArray<HistoryModel*>*) getCurrentWebViewHistories{
    return [self.wv__vc_paths objectForKey:[GlobalState getCurrentWebView]];
}
- (void) clearHistory:(XEngineWebView*) key{
    NSMutableArray *discardedItems = [NSMutableArray array];
    NSMutableArray* histories = [self.wv__vc_paths objectForKey:key];
    if(histories){
        for (HistoryModel *item in histories) {
            if (!item.vc)
                [discardedItems addObject:item];
        }
    }
    [histories  removeObjectsInArray:discardedItems];
}
- (void)addCurrentWebViewHistory:(HistoryModel *) history_model{
    XEngineWebView* key= [GlobalState getCurrentWebView];

    NSMutableArray* histories = [self.wv__vc_paths objectForKey:key];
    if(histories == nil ){
        histories=[NSMutableArray new];
        [self.wv__vc_paths setObject:histories forKey:key];
    }
    [histories addObject:history_model];
    [self clearHistory:key];
}


@end

File: modules/x-engine-native-jsi/android/native-jsi/src/main/java/com/zkty/nativ/jsi/GlobalState.java
package com.zkty.nativ.jsi;

import android.content.Context;

import com.zkty.nativ.jsi.webview.XEngineWebView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GlobalState {
    private XEngineWebView s_showing_webview;
    private Map<XEngineWebView, ArrayList<HistoryModel>> wv__vc_paths;

    private GlobalState() {
        wv__vc_paths = new HashMap<>();
    }

    private static volatile GlobalState instance = null;

    public static GlobalState sharedInstance() {
        if (instance == null) {
            synchronized (GlobalState.class) {
                if (instance == null) {
                    instance = new GlobalState();
                }
            }
        }
        return instance;
    }

    /**
     * Webview 初始化
     * EngineSdk初始化时调用
     */
    public void init(Context context) {
    }

    public String getLastHost() {
        if (s_showing_webview != null) {
            ArrayList<HistoryModel> histories = wv__vc_paths.get(s_showing_webview);
            if (histories != null && histories.size() > 0) {
                return histories.get(histories.size() - 1).host;
            }
        }
        return null;
    }

    public void setCurrentWebView(XEngineWebView webView) {
        this.s_showing_webview = webView;
    }

    public XEngineWebView getCurrentWebView() {
        return s_showing_webview;
    }

    public List<HistoryModel> getCurrentWebViewHistories() {
        if (s_showing_webview != null) {
            return wv__vc_paths.get(s_showing_webview);
        }
        return new ArrayList<>();
    }

    public void clearHistory(XEngineWebView key) {
        List<HistoryModel> discardedItems = new ArrayList<>();
        List<HistoryModel> histories = wv__vc_paths.get(key);
        if (histories != null) {
            for (HistoryModel item : histories) {
                if (!item.isResume)
                    discardedItems.add(item);
            }
            wv__vc_paths.get(key).removeAll(discardedItems);
        }
    }

    public void addCurrentWebViewHistory(HistoryModel model) {
        List<HistoryModel> histories = wv__vc_paths.get(s_showing_webview);
        if (histories==null){
            histories = new ArrayList<>();
        }
        histories.add(model);
    }
}
